### PR TITLE
Enable to sync registry creds by proxying CRI

### PR DESCRIPTION
**Pull request by ktock**
Fixes: https://github.com/containerd/stargz-snapshotter/issues/270

Though stargz snpashotter supports kubeconfig-based authentication on Kuberentes, this has some drawbacks including that it requires additional permission (listing/watching secrets) to the node.

For solving this issue, this commit introduces an alternative authentication method based on CRI proxy.

This allows stargz snapshotter to work as a proxy of CRI Image Service. Stargz Snapshotter exposes CRI Image Service API on its unix socket (i.e. `/run/containerd-stargz-grpc/containerd-stargz-grpc.sock`). It acquires registry creds by scanning requests on CRI PullImage API.

The following configuration enables this authentication mode.
`--image-service-endpoint=unix:///run/containerd-stargz-grpc/containerd-stargz-grpc.sock` option is needed to kubelet.

```toml
[cri_keychain]
enable_keychain = true
image_service_path = "/run/containerd/containerd.sock"
```

### CODE REVIEW

File: cmd/stargz-store/pool.go [hunk 1]
```diff
@@ -409,9 +410,14 @@ func (p *pool) writeManifestAndConfig(refspec reference.Spec, manifest ocispec.M
 	return mPath, cPath, nil
 }
 
-func fetchManifestAndConfig(ctx context.Context, hosts docker.RegistryHosts, refspec reference.Spec) (ocispec.Manifest, ocispec.Image, error) {
+func fetchManifestAndConfig(ctx context.Context, hosts source.RegistryHosts, refspec reference.Spec) (ocispec.Manifest, ocispec.Image, error) {
 	resolver := docker.NewResolver(docker.ResolverOptions{
-		Hosts: hosts,
+		Hosts: func(host string) ([]docker.RegistryHost, error) {
+			if host != refspec.Hostname() {
+				return nil, fmt.Errorf("unexpected host %q for imgae ref %q", host, refspec.String())
```
Comment (AkihiroSuda) [inline cmd/stargz-store/pool.go:417]: Typo: imgae

File: docs/overview.md [hunk 2]
```diff
@@ -118,6 +121,26 @@ Stargz snapshotter doesn't share credentials with containerd so credentials spec
...
 ```
 
+#### CRI-based authentication
+
+Following configuration enables stargz snapshotter to pull private images on Kubernetes.
+The snapshotter works as a proxy of CRI Image Service and exposes CRI Image Service API on the snapshotter's unix socket (i.e. `/run/containerd-stargz-grpc/containerd-stargz-grpc.sock`).
+The snapshotter acquires registry creds by scanning requests.
+
+You must specify `--image-service-endpoint=unix:///run/containerd-stargz-grpc/containerd-stargz-grpc.sock` option to kubelet.
+This is the recommended way to enable lazy pulling of private images on Kubernetes.
```
Comment (AkihiroSuda) [inline docs/overview.md:131]: I’m feeling kubeconfig-based auth is more preferable for practitioners, as they wouldn’t want to modify kubelet args.

Comment (AkihiroSuda) [reply docs/overview.md:131]: Also, depending on the deployment tool, you can’t specify the image service socket that is different from the runtime socket. (Does stargz snapshotter also proxify the runtime service?)

File: cmd/stargz-store/pool.go (hunk 1 above)
Comment (ktock) [reply cmd/stargz-store/pool.go]: Fixed it.

File: docs/overview.md (hunk 2 above)
Comment (ktock) [reply docs/overview.md]: Thank you for the comment. Fixed it.

**Review by AkihiroSuda — APPROVED** [review]

**Pull request merged**